docs/new_protocol.asciidoc

```diff
@@ -0,0 +1,463 @@
+models to follow. For UDP protocols, you can look at the DNS module.
+
+All protocol modules implement the `TcpProtocolPlugin` or the
+`UdpProtocolPlugin` (or both) from the following listing (found in
+`protos/protos.go`).
+
+[source,go]
+----------------------------------------------------------------------
+// Functions to be exported by a protocol plugin
+type ProtocolPlugin interface {
+	// Called to initialize the Plugin
+	Init(test_mode bool, results chan common.MapStr) error
+
+	// Called to return the configured ports
+	GetPorts() []int
+}
+
+type TcpProtocolPlugin interface {
+	ProtocolPlugin
+
+	// Called when TCP payload data is available for parsing.
+	Parse(pkt *Packet, tcptuple *common.TcpTuple,
+		dir uint8, private ProtocolData) ProtocolData
+
+	// Called when the FIN flag is seen in the TCP stream.
+	ReceivedFin(tcptuple *common.TcpTuple, dir uint8,
+		private ProtocolData) ProtocolData
+
+	// Called when a packets are missing from the tcp
+	// stream.
+	GapInStream(tcptuple *common.TcpTuple, dir uint8, nbytes int,
+		private ProtocolData) (priv ProtocolData, drop bool)
+}
+
+type UdpProtocolPlugin interface {
+	ProtocolPlugin
+
+	// ParseUdp is invoked when UDP payload data is available for parsing.
+	ParseUdp(pkt *Packet)
+}
+----------------------------------------------------------------------
+
+At the high level, the protocols plugins receive raw packet data via the
+`Parse()` or `ParseUdp()` methods and produce objects that will be indexed into
+Elasticsearch via the `results` channel.
+
+The `Parse()` and `ParseUdp()` methods are called for every packet that is
+sniffed and is using one of the ports of the protocol as defined by the
+`GetPorts()` function. They receive the packet in a Packet object, which looks
+like this:
+
+[source,go]
+----------------------------------------------------------------------
+type Packet struct {
+	Ts      time.Time <1>
+	Tuple   common.IpPortTuple <2>
+	Payload []byte <3>
+}
+----------------------------------------------------------------------
+
+<1> The timestamp of the packet
+<2> The source IP address, source port, destination IP address, destination port
+combination.
+<3> The application layer payload (i.e. without the IP and TCP/UDP headers) as a
+byte slice.
+
+The objects sent through the `results` channel have the type `common.MapStr`
+which is essentially a `map[string]interface{}` with a few more convenience
+https://github.com/elastic/libbeat/blob/fae9cf861b58f09cf578245e45415899f4151d32/common/mapstr.go[methods]
+added.
+
+Besides the `Parse()` function, the TCP layer also calls the `ReceivedFin()`
+when a TCP stream is closed and the `GapInStream()` functions when packet loss
+is detected in a TCP stream. The protocol module can use these callbacks to take
+decisions about what to do with partial data received. For example, for the
+HTTP/1.0 protocol, the end of connection is used to know when the message is
+finished.
+
```

Comment:
Using ReceivedFin can be tricky. for example:
- on tcp shutdown 2 fins are received. One for each direction
- ReceivedFin only called if packet seen (not always the case)
- no signal/callback if tcp thread times out
=> if application layer connection data, messages or transactions are in 'global' map a timer to cleanup resources is required